### DISCUSSION TITLE
How can I inject a CDI bean in a custom ConfigSource or access Http request context in a custom ConfigSource ?

### DISCUSSION THREAD

**Discussion by omasseau** [quarkusio/quarkus#41904] [Q&A] [answered]
Hello,

I need to implement a [custom ConfigSource](https://quarkus.io/guides/config-extending-support#custom-config-source) to fetch the value of a property dynamically at runtime in my REST service.
The logic that I need to implement to fecth the correct property value is dependent on the hostname used to call the REST service.

Currently to know the hostname I read the injected UriInfo in my REST service: 

```
    @Inject
    RequestHostBean requestHostBean; //This is a @RequestScoped bean that just stores the request host name

    @GET
    public RestResponse<Void> get(@Context UriInfo uriInfo) {
    	String host = uriInfo.getBaseUri().getHost();
       	requestHostBean.setHost(host);       
      	...
      	String hostDependentPropertyValue = ConfigProvider.getConfig().getConfigValue("hostDependentProperty").getValue();
        ...       
      	return RestResponse.ok();
    }
```

I need to access this host (either by getting the RequestHostBean instance or UriInfo) in the implementation of my custom ConfigSource (or [ConfigSourceFactory](https://quarkus.io/guides/config-extending-support#config-source-factory))

How is this possible as a ConfigSource/ConfigSourceFactory is not a managed CDI bean. Is there any easy way ?

[a comment by an automated account is not shown]

**Comment by manovotn** [comment, marked as answer]
> How is this possible as a ConfigSource/ConfigSourceFactory is not a managed CDI bean. Is there any easy way ?

If you want to resolve a CDI from an arbitrary non-CDI location you can use programmatic lookup.
The Quarkus way of doing this is as follows:

```java
RequestHostBean myBean = Arc.container().instance(RequestHostBean.class).get();
```

The CDI spec way is through `CDI.current().select(...).get()`.

Either way, the bean is only really "ready" in runtime, attempting to access it earlier will result in an error.
Therefore, I am not sure this is what you are looking for with config source?

**Reply by omasseau** [reply]
This works perfectly thanks.
This is exactly what I was looking for. I only need to fetch the properties at runtime when a call is made to a rest service.

Now I am being able to do something like this : 

```
@RequestScoped
@Path("/test")
public class TestResource {

    // Value loaded dynamically through CustomConfigSource
    // For injected value to be determined on each request, the class must be request scoped, else we have to use ConfigProvider.getConfig().getValue() if the class is application scoped
    @ConfigProperty(name="test")
    String testValue;

    @GET
    public void test() {
        Log.info("Test property injected dynamically on each request : " + testValue);
    }
}
```

**Reply by radcortez** [reply]
Yes, I usually recommend not mixing Sources and CDI beans. Sources are required before and after the CDI container (which also uses the Config system to configure itself), so relying on CDI beans is not a good idea.

There are some tricks to make it work, but what I usually observe is that they will create additional issues. For instance, https://github.com/quarkusio/quarkus/issues/41313 is a good example.

If you need to retrieve configuration dynamically by the hostname, maybe you could just set a `Map<String, ?>` configuration and then query the configuration map by key using the hostname. This is what we do with datasources: https://github.com/quarkusio/quarkus/blob/39bb23f5d432b6b2b6a547e0127c861672f743db/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourcesJdbcRuntimeConfig.java
